cpp/src/arrow/dataset/file_parquet.cc

```diff
@@ -67,8 +72,28 @@ parquet::ReaderProperties MakeReaderProperties(
     properties.disable_buffered_stream();
   }
   properties.set_buffer_size(parquet_scan_options->reader_properties->buffer_size());
+
+#ifdef PARQUET_REQUIRE_ENCRYPTION
+  auto parquet_decrypt_config = parquet_scan_options->parquet_decryption_config;
+
+  if (parquet_decrypt_config != nullptr) {
+    auto file_decryption_prop =
+        parquet_decrypt_config->crypto_factory->GetFileDecryptionProperties(
+            *parquet_decrypt_config->kms_connection_config,
+            *parquet_decrypt_config->decryption_config, path, filesystem);
+
+    parquet_scan_options->reader_properties->file_decryption_properties(
+        std::move(file_decryption_prop));
+  }
+#else
+  if (parquet_scan_options->parquet_decryption_config != nullptr) {
+    return Status::NotImplemented("Encryption is not supported in this build.");
+  }
```

Comment:
Ouch, my bad. We cannot return a `Status` here, we must instead throw a Parquet exception.
(this can be seen on some CI jobs that have encryption disabled: 
https://github.com/apache/arrow/actions/runs/6411553066/job/17408394537?pr=34616#step:5:3088 )

Perhaps something like (untested):
```suggestion
  if (parquet_scan_options->parquet_decryption_config != nullptr) {
    parquet::ParquetException::NYI("Encryption is not supported in this build.");
  }
```